### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by urban48**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,77 @@
+nose-ittr
+=========
+nose extension for supporting parametrized testing.
+---------------------------------------------------
```
Comment (talbenbasat) [inline README.rst:4]: add . at  the end of the sentence (line 5)

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        TestFoo.test_even_2 ... .ok
+        TestFoo.test_even_3 ... FAIL
+        TestFoo.test_even_4 ... .ok
+        TestFoo.test_no_remainder_2_6 ... .ok
+        TestFoo.test_no_remainder_2_15 ... FAIL
+        TestFoo.test_no_remainder_3_6 ... .ok
+        TestFoo.test_no_remainder_3_15 ... .ok
+
+
+**notes:**
```
Comment (talbenbasat) [inline README.rst]: . at the end of the sentence

File: nose_ittr/Ittr_multiplayer.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    return update_attr
+
+
+class IttrMultiplayer(type):
+    """
+    Multiples the tests in a given test class by
+    the parameters given in ittr decorator.
+    """
+    def __new__(mcs, name, bases, dct):
+
```
Comment (talbenbasat) [inline nose_ittr/Ittr_multiplayer.py]: attributeName is a not PEP8

File: nose_ittr/Ittr_multiplayer.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                suffix = re.sub(r'\W+', '',
+                                str(prod.values())
+                                .translate(None, "[]'")
+                                .replace(',', '_'))
+                logging.debug('method suffix: {0}'.format(suffix))
+                # in case itts passed are empty
+                if not suffix:
+                    continue
+                new_func_name = attributeName + '_' + suffix
+
```
Comment (talbenbasat) [inline nose_ittr/Ittr_multiplayer.py]: why @ in the comment

File: nose_ittr/Ittr_multiplayer.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            for prod in products:
+                logging.debug('method product: {0}'.format(prod))
+                suffix = re.sub(r'\W+', '',
+                                str(prod.values())
+                                .translate(None, "[]'")
+                                .replace(',', '_'))
+                logging.debug('method suffix: {0}'.format(suffix))
+                # in case itts passed are empty
+                if not suffix:
+                    continue
```
Comment (talbenbasat) [inline nose_ittr/Ittr_multiplayer.py]: add log

File: nose_ittr/Ittr_multiplayer.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            b = [map(lambda value: (key, value), values)
+                 for key, values in attribute.ittr.iteritems() if values]
+            products = map(dict, product(*b))
+
+            for prod in products:
+                logging.debug('method product: {0}'.format(prod))
+                suffix = re.sub(r'\W+', '',
+                                str(prod.values())
+                                .translate(None, "[]'")
+                                .replace(',', '_'))
```
Comment (talbenbasat) [inline nose_ittr/Ittr_multiplayer.py]: move the logging to be after the continue  
and please add a breakline instead

File: nose_ittr/Ittr_multiplayer.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+                # combine both product and @ittr dict to be added to new method
+                func_params = dict(attribute.func_dict, **prod)
+                mirror_func = mcs._attribute_injector(attribute, **func_params)
+                setattr(mirror_func, 'ittr', prod)
+
+                # assign new name and docstring and save back at our class
+                mirror_func.func_name = new_func_name
+                mirror_func.func_doc = attribute.func_doc
+                dct[new_func_name] = mirror_func
```
Comment (talbenbasat) [inline nose_ittr/Ittr_multiplayer.py]: breakline

File: nose_ittr/Ittr_multiplayer.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            # transfer ittr and attr to self when called
+            self = fargs[0]
+            for name, value in keywords.iteritems():
+                setattr(self, name, value)
+            return func(*fargs, **fkeywords)
+
+        # transfers all attr and ittr to newfunc
+        for name, value in keywords.iteritems():
+            setattr(wrapper, name, value)
+        setattr(wrapper, 'keywords', keywords)
```
Comment (talbenbasat) [inline nose_ittr/Ittr_multiplayer.py]: remove unneeded comment
add breakline instead

File: nose_ittr/Ittr_multiplayer.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                # assign new name and docstring and save back at our class
+                mirror_func.func_name = new_func_name
+                mirror_func.func_doc = attribute.func_doc
+                dct[new_func_name] = mirror_func
+            # set no test flag to original test method
+            attribute.__test__ = False
+        return type.__new__(mcs, name, bases, dct)
+
+    @classmethod
+    def _attribute_injector(cls, func, **keywords):
```
Comment (talbenbasat) [inline nose_ittr/Ittr_multiplayer.py]: change the name of wrapper to inject or other informative

File: nose_ittr/test/test_ittr_multiplayer.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    def test_correct_mirroring(self):
+        """
+        Test metaclass set ittr attributes in mirrors functions
+        """
+        ok_(hasattr(self.test_class.test_method_8_val_a_val_b_val_c,
+                    'attr_one'))
+        ok_(hasattr(self.test_class.test_method_8_val_a_val_b_val_c,
+                    'attr_two'))
+        ok_(hasattr(self.test_class.test_method_8_val_a_val_b_val_c,
+                    'attr_three'))
```
Comment (talbenbasat) [inline nose_ittr/test/test_ittr_multiplayer.py:29]: breakline

File: nose_ittr/test/test_ittr_multiplayer.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    def test_ittr_attribute_insertion(self):
+        """
+        Tests metaclass adds ittr attribute to test
+        """
+        assert_in('attr_one',
+                  self.test_class.test_method_8_val_a_val_b_val_c.ittr)
+        assert_in('attr_two',
+                  self.test_class.test_method_8_val_a_val_b_val_c.ittr)
+        assert_in('attr_three',
+                  self.test_class.test_method_8_val_a_val_b_val_c.ittr)
```
Comment (talbenbasat) [inline nose_ittr/test/test_ittr_multiplayer.py:41]: breakline

File: nose_ittr/test/test_ittr_multiplayer.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+
+    @ittr(attr_one=['val_a', 'val_b'], attr_two=['val_c'])
+    @attr(id=1234, section='test')
+    def test_method_7(self):
+        pass
+
+    @ittr(attr_one=['val_a'], attr_two=['val_b'])
+    @ittr(attr_three=['val_c', 'val_d'])
+    def test_method_8(self):
```
Comment (talbenbasat) [inline nose_ittr/test/test_ittr_multiplayer.py:137]: make docstring upper case in the first letter

File: setup.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    for filename in filenames:
+        with io.open(filename, encoding=encoding) as f:
+            buf.append(f.read())
+    return sep.join(buf)
+
+long_description = read('README.rst', 'CHANGES.txt')
+
+
+setup(
+    name='nose_ittr',
```
Comment (talbenbasat) [inline setup.py:32]: 0.0.1

File: setup.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    version='1.0',
+    url='https://github.com/taykey/nose-ittr',
+    license='Apache Software License',
+    author='Sergey Ragatsky',
+    author_email = "serg@taykey.com",
+    tests_require=['nose'],
+    install_requires=[''],
```
Comment (talbenbasat) [inline setup.py:39]: nose extension

File: setup.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    description='nose expansion for supporting parametrized testing',
+    long_description=long_description,
+    packages=find_packages(),
+    include_package_data=False,
+    platforms='any',
+    keywords='nose nosetest docstring metaclass parametrized testing',
+    test_suite='nose.collector',
```
Comment (talbenbasat) [inline setup.py:46]: add contributors

File: README.rst [hunk 16]
```diff
@@ -0,0 +1,77 @@
+nose-ittr
+=========
+nose extension for supporting parametrized testing.
+---------------------------------------------------
+Allow developer to run the same test over and over again using different values.
```
Comment (MarounMaroun) [inline README.rst]: Better: Allows (note the "s"), and remove the dot after "values" (Agreed with Tal)

File: README.rst [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+Main Features:
+ * Very easy to integrate with existing tests.
```
Comment (MarounMaroun) [inline README.rst]: Dots shouldn't appear after bullets

File: README.rst [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+            pass
+        
+        def teardown(self):
+            pass
+            
+        @ittr(number=[1, 2, 3, 4])
+        def test_even(self):
+            assert_equal(self.number % 2, 0)
+            
+
```
Comment (MarounMaroun) [inline README.rst:40]: PEP8 (sorry, it applies also to READMEs)

File: README.rst (hunk 2 above)
Comment (MarounMaroun) [inline README.rst]: "Notes:" with capital "N"

File: nose_ittr/Ittr_multiplayer.py [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+            # create product of all the iterators
+            b = [map(lambda value: (key, value), values)
+                 for key, values in attribute.ittr.iteritems() if values]
+            products = map(dict, product(*b))
+
+            for prod in products:
+                logging.debug('method product: {0}'.format(prod))
+                suffix = re.sub(r'\W+', '',
+                                str(prod.values())
```
Comment (MarounMaroun) [inline nose_ittr/Ittr_multiplayer.py]: Maybe replaced string should be generated one line above, for clarity

**Pull request merged**